File: baekjoon-Algorithm/algorithm-middle/dynamic-programming/N1965.py
import sys

input = sys.stdin.readline

n = int(input())
nlist = list(map(int, input().split()))

dp = [1]

for i in range(1, n):
    d = []

    for j in range(i):
        if nlist[j] < nlist[i]:
            d.append(dp[j])

    if not d:
        dp.append(1)
    else:
        dp.append(max(d) + 1)


print(max(dp))

File: baekjoon-Algorithm/algorithm-middle/dynamic-programming/N1965.java
import java.io.*;
import java.util.*;

public class N1965 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        int[] nlist = new int[n];

        StringTokenizer st = new StringTokenizer(br.readLine(), " ");

        for (int i= 0; i < n; i++) {
            nlist[i] = Integer.parseInt(st.nextToken());
        }

        int[] dp = new int[n];
        dp[0] = 1;
        int max = Integer.MIN_VALUE;

        for (int i = 1; i < n; i++) {
            dp[i] = 1;
            for (int j = 0; j < i; j++) {
                if (nlist[j] < nlist[i]){
                    dp[i] = Math.max(dp[j] + 1, dp[i]);
                }
            }
            max = Math.max(dp[i], max);
        }

        System.out.println(max);
    }
}
